Simplify bulk request execution

areek opened the pull request:
Currently, bulk item requests can be any ActionRequest, this PR
restricts bulk item requests to DocumentRequest. This simplifies
handling failures during bulk requests. Additionally, a new enum
is added to DocumentRequest to represent the intended operation
to be performed by a document request (`create`, `index`, `update` 
and `delete`), which was previously represented with a mix of strings 
and index request operation type.
Now, index request operation type reuses the new enum to specify 
whether the request should `create` or `index` a document.
Restricting bulk requests to DocumentRequest further simplifies
execution of shard-level bulk operations to use the same failure 
handling for index, delete and update operations.
This PR also fixes a bug which executed delete operations twice for 
replica copies while executing bulk requests.

relates https://github.com/elastic/elasticsearch/pull/19105

bleskes commented on core/src/main/java/org/elasticsearch/action/bulk/BulkItemResponse.java:
```diff
@@ -322,7 +328,7 @@ public void readFrom(StreamInput in) throws IOException {
     public void writeTo(StreamOutput out) throws IOException {
         out.writeVInt(id);
-        out.writeString(opType);
+        out.writeByte(opType.getId());
```
we need bwc here too

areek replied:
bwc added

areek commented on core/src/main/java/org/elasticsearch/action/DocumentRequest.java:
```diff
@@ -70,4 +72,78 @@
+
+        OpType(int op) {
+            this.op = (byte) op;
+            this.lowercase = this.toString().toLowerCase(Locale.ENGLISH);
```
changed it to ROOT, but there are a few places (e.g. `DocWriteResponse`) where we use ENGLISH too. maybe we should change it to ROOT as well?

areek commented on core/src/main/java/org/elasticsearch/action/bulk/BulkRequest.java:
```diff
@@ -577,15 +574,17 @@ public void writeTo(StreamOutput out) throws IOException {
         out.writeVInt(requests.size());
-        for (ActionRequest<?> request : requests) {
+        for (DocumentRequest<?> request : requests) {
             if (request instanceof IndexRequest) {
```
done

areek commented on core/src/main/java/org/elasticsearch/action/bulk/TransportShardBulkAction.java:
```diff
@@ -132,203 +126,84 @@ protected boolean resolveIndex() {
-            location = locationToSync(location, result.getLocation());
+            WriteResult<? extends DocWriteResponse> writeResult = innerExecuteBulkItemRequest(metaData, indexShard,
+                request, requestIndex);
+            if (writeResult.getResponse().getResult() != DocWriteResponse.Result.NOOP) {
```
I wanted to be more strict as only a noop result can be valid with a null location. changed it to be an assert instead.

areek commented on core/src/main/java/org/elasticsearch/action/bulk/TransportShardBulkAction.java:
```diff
@@ -132,203 +126,84 @@ protected boolean resolveIndex() {
-                setResponse(item, item.getPrimaryResponse());
+            BulkItemRequest item = request.items()[requestIndex];
+            DocumentRequest<?> documentRequest = item.request();
+            if (ExceptionsHelper.status(e) == RestStatus.CONFLICT) {
```
fixed

areek commented:
Thanks @bleskes for the review :). I addressed all the minor comments. 

> I also think we can potentially simplify further and make the DocumentRequest inherit from WriteReplicationRequest (and call DocWriteRequest)

I really like the idea of making DocumentRequest inherit from WriteReplicationRequest. While giving this a go, I refactored update operation to be a replication operation (updates are a DocumentWriteRequest but does't use the replication operation instead delegates to index and delete operations for replication). This cleans up the dedicated transport update action (`TransportInstanceSingleOperationAction`) in favour of reusing replication action and duplicate code for update retry logic and translation. The one big downside is it would be difficult to implement wire bwc with 5.0. WDYT?

bleskes commented on core/src/main/java/org/elasticsearch/action/bulk/TransportShardBulkAction.java:
```diff
@@ -338,105 +214,56 @@ private void setResponse(BulkItemRequest request, BulkItemResponse response) {
-                    return new UpdateResult(translate, indexRequest, retry, cause, null);
+                } else {
+                    assert translate.getResponseResult() == DocWriteResponse.Result.UPDATED;
+                    update.setGetResult(updateHelper.extractGetResult(updateRequest, updateRequest.concreteIndex(), indexResponse.getVersion(), translate.updatedSourceAsMap(), translate.updateSourceContentType(), indexSourceAsBytes));
```
why do we now return always the get result on updates? it seems to be different [before](https://github.com/elastic/elasticsearch/blob/master/core/src/main/java/org/elasticsearch/action/bulk/TransportShardBulkAction.java#L251)

areek replied:
I changed it to match the logic in [TransportUpdateAction.shardOperation](https://github.com/elastic/elasticsearch/blob/master/core/src/main/java/org/elasticsearch/action/update/TransportUpdateAction.java#L175) but that caused CI failures, I reverted the change to be the same as update operation in bulk

bleskes commented:
It also seems CI isn't happy.

areek commented:
Thanks @bleskes for the feedback. I updated the PR, addressing your comments and the CI is happy. Could you take a look

bleskes commented:
LGTM. Awesome @areek
